x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/blob/BlobStoreCacheMaintenanceService.java

```diff
@@ -236,4 +353,285 @@ public void onFailure(Exception e) {
+                            maintenanceTask.searchResponse = response;
+                            maintenanceTask.searchAfter = null;
+                            executeNext(maintenanceTask);
+                        }
+
+                        @Override
+                        public void onFailure(Exception e) {
+                            complete(maintenanceTask, e);
+                        }
+                    });
+                    return;
+                }
+
+                final SearchHit[] searchHits = searchResponse.getHits().getHits();
+                if (searchHits != null && searchHits.length > 0) {
+                    final ClusterState state = clusterService.state();
+                    final BulkRequest bulkRequest = new BulkRequest();
+
+                    final Set<Tuple<String, String>> knownSnapshots = new HashSet<>();
+                    final Set<Tuple<String, String>> missingSnapshots = new HashSet<>();
+                    final Set<String> knownRepositories = state.metadata()
+                        .custom(RepositoriesMetadata.TYPE, RepositoriesMetadata.EMPTY)
+                        .repositories()
+                        .stream()
+                        .map(RepositoryMetadata::name)
+                        .collect(Collectors.toSet());
+
+                    Object[] lastSortValues = null;
+                    for (SearchHit searchHit : searchHits) {
+                        assert searchHit.getId() != null;
+                        assert searchHit.hasSource();
+
+                        // See {@link BlobStoreCacheService#generateId}
+                        // doc id = {repository name}/{snapshot id}/{snapshot index id}/{shard id}/{file name}/@{file offset}
+                        final String[] parts = Objects.requireNonNull(searchHit.getId()).split("/");
+                        assert parts.length == 6 : Arrays.toString(parts) + " vs " + searchHit.getId();
+
+                        boolean delete = false;
+                        lastSortValues = searchHit.getSortValues();
+
+                        final String repositoryName = parts[0];
+                        if (knownRepositories.contains(repositoryName) == false) {
+                            logger.trace(
+                                "deleting blob store cache entry [id:{}, repository:{}, reason: repository does not exist]",
+                                searchHit.getId(),
+                                repositoryName
+                            );
+                            delete = true;
+                        } else {
+                            final Tuple<String, String> snapshot = Tuple.tuple(parts[1], parts[2]);
+                            boolean isMissing = missingSnapshots.contains(snapshot);
+                            boolean isKnown = knownSnapshots.contains(snapshot);
+                            if (isMissing
```

Comment:
Rather than maintain this extra state and loop through the indices for every snapshot we see, I wonder if we could simply find the set of snapshots to keep after opening the PIT? Since the PIT will not return docs indexed after it was opened, that should be relatively safe.

There is a race condition though, in that we risk a searchable snapshot in the process of being mounted but this node does not know about it yet. That race also exists with the current structure.

I can think of a few options:

1. Remember max seq-no from last round and only allow deleting entries with lower seq-no than that.
2. Wait 10 sec after opening PIT before proceeding.
3. Collect the snapshot ids to delete but wait a round before deleting, rechecking that they can still be deleted.
4. Ignore the issue, put in a comment and accept that we rarely delete a bit too eagerly.

Do you have other ideas/options?

Reply:
Thanks for this feedback. I used to consider this cache as a best effort thing so in my mind I accepted that we sometimes delete cache entries too early. With the fact that it runs not so frequently and that the cluster state is "refreshed" for every batch of docs to scan I found it acceptable.

But I do like your suggestion of using a timestamp. It fits well with the current behavior (assuming we list the snapshots to keep once the PIT is opened) and in fact I already added a `creation_time` field to the docs as I anticipated it would be useful in the future... for clean up.

So I'd go with listing the snapshots to keep and use the existing timestamp.

Reply:
Sounds great, I somehow missed the timestamp (was not given as args to `putAsync`), but do see it now. I agree to this path forward.

Reply:
> I somehow missed the timestamp (was not given as args to `putAsync`)

I moved the creation time as an argument of `putAsync` in 6ac5a6e344e6b3dc63fab05d1675b934c50ff132. This is more natural and it is useful in tests too.

I pushed 70b7666719620aaf0e0feac645bf4f9b9f604d57 to compute the existing searchable snapshots before the first bulk request is executed. I updated the test to check that recent cache entries are correctly ignored.